### ISSUE TITLE
Is XACML the right name?

### ISSUE THREAD

**Issue by steven-legg** [oasis-tcs/xacml-spec#35]
We're intending to make it possible to implement the next version of XACML exclusively using JSON (so without touching XML at all). However, anyone seeing "XACML 4.0" is likely to assume that it is still based in XML and may not look any deeper. Should we be calling the next iteration something else?

What we call things is intertwined with how we separate the specification into multiple documents and how we present the JSON (Schema) versus XML (Schema) representations, which we give prominence and what is deemed optional. We should consider a potential rename first because it will drive those decisions.

Here are some thoughts in no particular order.

The X in XACML doesn't really stand for XML but kind of implies it, so we swap that out (JACML) or drop it and be agnostic (ACML). The "ML" (markup language) is also a bit of a throwback to the XML world so maybe we just shorten JACML to JACL.

A question arises over the version number. JACL 4.0 and ACML 4.0 are a bit strange without there being a preceding JACL 3 or ACML 3.

We could go with an overarching name (e.g., ACML 1.0) that has two representations known as JACL 1.0 and XACML 4.0, either of which could be an optional, separate profile. That solves the versioning question and indicates the continuity from XACML 3.0 to XACML 4.0.

We could name the representations relative to the overarching name, e.g., ACML-J 1.0, ACML-X 1.0 (a.k.a. XACML 4.0). Or ACML-JSON and ACML-XML.

ACML still has that "M" stuck in there, but ACL is a bit too nondescript. Perhaps we swap the X to E and drop the M: XACML -> EACL, perhaps with representations JEACL 1.0 and XACML 4.0.

**Comment by steven-legg** [comment]
ALFA 2.0 can slot in beside JACL 1.0 and XACML 4.0.

**Comment by cdanger** [comment]
I would avoid anything with suffix "ACL" which may lead people to think this is a language for ACLs = Access Control Lists. (XACML is much more, this is ABAC!)
Maybe **ACPL** instead? (P for Policy), so JACPL, XACPL (a.k.a. XACML), YACPL (for YAML)...

**Comment by steven-legg** [comment]
If we swap "access control" (AC) with "authorization" (A) we get the easier to pronounce JAPL, XAPL and YAPL ("japple", "zapple" and "yapple") and of course the specification that ties them together is the APL Core :-).

**Comment by humantypo** [comment]
JAzL, XAzL? AuthZ Language. Mixed cases, nice syllabic tempo. Rhymes with dazzle 😄

**Comment by cdanger** [comment]
> If we swap "access control" (AC) with "authorization" (A) we get the easier to pronounce JAPL, XAPL and YAPL ("japple", "zapple" and "yapple") and of course the specification that ties them together is the APL Core 😄.

Yeah I thought of  "APL" as well (as in Abac Policy Language or Authorization Policy Language works too) but then I stumbled upon an open source ABAC project called APL:  https://github.com/intuit/identity-authz-apl
... which decided to go with its own language, so not XACML :-( .

But If you think that's no big deal, I'm fine with either APL or AzL (JAzL sounds cool for sure, but XAzL a bit hard to pronounce for me).

**Comment by steven-legg** [comment]
Another suggestion to kick around. OASIS Authorization Policy Language (OAPL, sounds like OPAL), which leads to JOAPL (JOPAL) and YOAPL (YOPAL). The XML version is XACML 4.0.

**Comment by cdanger** [comment]
I think we should set up a vote online for TC members to vote on the proposals we have so far at some point (I can do it with google forms for instance):
- AzL (Authorization Language): JAzL, YAzL
- AzPL (Authorization Policy Language): JAzPL, YAzPL
- OAPL (OASIS Authorization Policy Language): JOAPL, YOAPL

(I've just added the second one.) In all cases, we should keep the name "XACML" for the XML variant (at least in 4.0), that's why I don't mention it above, this is implicit.  Let me know what you think.

**Comment by humantypo** [comment]
I just realized that I am going to have a hard time requesting Starter Documents for a specification *that doesn't have a name*...especially Core 😄 

I was also thinking that since we are looking at the Core specification as being agnostic linguistically and applicably we might want to consider something more general for that and then introduce the "language" (XML, JSON) and "scope" (Access Control, etc.) into the *Profile* names.

Perhaps something like SAR for Core (simple reference to the key elements) or SARC (SAR in Context)...?

**Comment by steven-legg** [comment]
> I just realized that I am going to have a hard time requesting Starter Documents for a specification _that doesn't have a name_...

Indeed.

What does SAR stand for?

I assume we can fiddle the title in the text anytime, but we can't readily change the URLs and file name. A number of our candidate names revolve around "PL", for policy language. If we start with that then the URLs and filenames won't be that incongruous with whatever overarching name we finally settle on. I'm thinking something like:
* Policy Language Version 1.0 - Core; https://docs.oasis-open.org/xacml/pl-v1.0-core/v1.0/wd01/pl-v1.0-core-wd01.htm
* Policy Language 1.0, JSON Syntax Version 1.0; https://docs.oasis-open.org/xacml/pl-1.0-json-1.0/v1.0/wd01/pl-1.0-json-v1.0-wd01.htm
* Policy Language 1.0, JSONPath Profile Version 1.0; https://docs.oasis-open.org/xacml/pl-1.0-json-path/v1.0/wd01/pl-1.0-json-path-v1.0-wd01.htm
* Policy Language 1.0, eXtensible Access Control Markup Language Version 4.0; https://docs.oasis-open.org/xacml/pl-1.0-xacml-4.0/v4.0/wd01/pl-1.0-xacml-v4.0-wd01.htm
* Policy Language 1.0, XPath Profile Version 1.0; https://docs.oasis-open.org/xacml/pl-1.0-xpath-1.0/v1.0/wd01/pl-1.0-xpath-v1.0-wd01.htm

I assume the first "xacml" in the URLs is determined by the TC name.

"Concepts" would be a potential substitute for "Core".

**Comment by steven-legg** [comment]
From the minutes:
> There was a general discussion on GitHub branch naming with the outcome being that the TC's current work would be in a branch and whatever was considered ready for Committee Specification would be merged with *main* so as to make consumption of the specification easier and to address the dynamic nature of many files creating the potential for a lot of "administrative churn". This was determined to be the best solution for pivoting away from the previous "Editor Submission" model that was used previously.

Thinking further on what we discussed in the TC meeting, it's not really going to work for the long term with all documents in the same repository. We can manage it with the five successors to the core spec for a while but their life cycles will diverge. Especially in the later stages the XPath and JSONPath profiles won't need as many revisions but we're sort of locking ourselves into revving the draft number just because the core needs to. Add in the other profiles like RBAC, Hierarchy, SAML, etc., some of which might not need to revert to wd-01, and it gets messy. If each document is in its own repository then the problem goes away.

The new TC rules let us treat committee specification drafts much like working drafts so we can rev the numbers more frequently without reviews and voting. This will be easier if each document is in its own repository. And we can still release working drafts for a document ahead of its first committee specification draft for interesting milestones.

When we do increment a working draft or committee specification draft we could rename (i.e. renumber) the existing file or freeze the existing file (for historical reference) and create a copy with the new number. I'm thinking we freeze and copy and leave out the older versions when we package things up for TC Admin to publish.

**Comment by cdanger** [comment]
I agree the profiles will have different life cycles from the core, all the more as we make the core spec more format-agnostic, therefore less subject to evolutions of XML, JSON, YAML standards (and possibly new formats in the future like CBOR for instance); and therefore would be better managed with separate github repositories. Also if you need to review the history of changes to the core document only, or to a specific profile, it seems difficult if everything is merged to the `main` branch in the end.

As an example, OpenC2 TC (OASIS) seems to follow [this approach](https://groups.oasis-open.org/communities/tc-community-home2?CommunityKey=a34c9baf-48b2-44c5-a567-018dc7d32296#tc-tools) (certain profiles have dedicated repositories).

**Comment by humantypo** [comment]
Yes, it really does make the CS look like a Working Draft. Challenging. Particularly from the consumption side. The OpenC2 repository makes a lot of sense in terms of managing updates for highly independent work artifacts. Assume that each of these was one of our Profiles (and XML, JSON and Core) and consider what the consumer friendly way to pull down all of the *compatible* versions into something that is a holistic spec would be?

Technically, I don't think the fact that there are multiple document "components" in a repository is the issue. Consider an app releasing with multiple libraries, all being updated on different tempos. The overall release version is what allows consumers to know what the latest "prod" version is for the *application*. Keep in mind that historically there were periods where the Core Specification ran ahead of some of the Profiles so grabbing the "main" branch of each may not guarantee compatibility. I can envision version management becoming even more complex as we are basically creating a "forked" specification with XML and JSON (where, say, the XML and JSON RBAC Profiles are not in lockstep on revisions...or will we need to make that mandatory?)

That said, I think the fact that Profiles are optional (outside of the new concept that at least one "language" Profile must be chosen) makes for a compelling case to keep Profiles separate at some level but, I am concerned that too much granularity will make consumption difficult.

How about we consider 3 "structural" repositories: **Core**, **XML** and **JSON**, with the latter two containing all *applicable* "traditional" Profiles? We then internally develop a versioning model that *spans* all 3 repositories and that the main branch in each represents the latest "approved" version (merged from working branch once OK'd by TC)? The goal here is that if a consumer pulls (main) Core + XML or Core + JSON, that the resultant specification will be consistent across all *included* documents.

Either that or it seems like we may need to provide either some sort of roadmap document that explains what combination of repository versions work together as a "Master Specification" or create and post a static file that the TC builds at each approved "milestone"...? 

P.S. All of this makes me wonder about calling the XML and JSON components "Profiles", given they are kinda "Co-Specs" on some level.

P.P.S I think I am starting to overthink this, so stopping for a sanity check. 🤯

**Comment by humantypo** [comment]
What does SAR stand for?

Subject, Action, Resource. I was playing around with what it is, rather than what it does, given our discussion our potential forking...

**Comment by steven-legg** [comment]
> When we do increment a working draft or committee specification draft we could rename (i.e. renumber) the existing file or freeze the existing file (for historical reference) and create a copy with the new number. I'm thinking we freeze and copy and leave out the older versions when we package things up for TC Admin to publish.

It looks like OpenC2 don't use use draft numbers in the file names in GitHub and only apply them when publishing a draft to the TC documents library. On the OASIS side it looks like business as usual.

**Comment by humantypo** [comment]
That makes sense. Using the names to store version information has been a construct that allows the traditional static library to act as a repository. Now that the docs are in an actual library it seems to me that the only version reference that is meaningful is that of the CS/Spec and that revisions be handled (opaquely to the Oasis library) in GitHub only. An issue that comes up there is that this means the only place to retrieve a CS/Spec is from the library, as git pulls will vary over time. 

On the plus side, business as usual has it’s advantages for consumers.

**Comment by humantypo** [comment]
Testing the "cc" mechanism to the TC list.

**Comment by cdanger** [comment]
Next step: summarize the options in the table with columns 

1.     AzL (Authorization Language): JAzL, YAzL
2.     AzPL (Authorization Policy Language): JAzPL, YAzPL
3.     OAPL (OASIS Authorization Policy Language): JOAPL, YOAPL

Send a message on the xacml-comments and xacml-users mailing lists with these options, explaining the meaning and motivation behind each of them.
Maybe organise a Google poll where people pick them in order of preference. Also explain that the XML version will still be called XACML for all options, so no name change there.

**Comment by humantypo** [comment]
I spent some time exploring names with the LLMs to see if I could come up with a name for the core specification and this one started to grow on me:

**LAPIS: Language-Agnostic Policy Interchange Specification**

My thinking was to find something that was “fundamental” in terms of what the specification is and its role in the access control process (which it is not mechanically limited to). I asked chatGPT to opine on it and this was some of its feedback:
```
LAPIS is an excellent acronym choice—it’s:
  * Pronounceable (like “lapis lazuli”),
  * Has an abstract, polished feel, and
  * Suggests something foundational or elemental.

Blurb:
LAPIS provides a structured, interoperable representation of access control policies that enables safe and consistent interchange across systems and policy engines, regardless of encoding format.

Intro Paragraph:

The Language-Agnostic Policy Interchange Specification (LAPIS) defines a platform-neutral representation for the definition and exchange of access control policies. Its purpose is to ensure that policies can be authored, validated, and transferred between systems in a consistent, semantically aligned manner. LAPIS supports multiple encoding syntaxes through well-defined mappings to XML (LAPIS-X), JSON (LAPIS-J), YAML (LAPIS-Y), and other future formats, facilitating tooling and automation without format lock-in.
```
This brings me back to the conundrum of naming the various output formats. I agree that we need to find a way to retain “XACML”, but at the same time I don’t think we want to create a bag of names due to the risk of confusion. Building upon the output above I was thinking that maybe a way to do this would be to create a base set of names and then provide a “nickname” for the XML version (something that is common in things like Linux versioning). In that situation the spec naming could look something like this:

Specification: LAPIS, v1.0
XML encoding: LAPIS-X v1.0 (XACML4)
JSON encoding: LAPIS-J, v1.0
YAML encoding: LAPIS-Y, v1.0
…
ALFA encoding: LAPIS-A, v1.0 (ALFA2) ...?

As discussed earlier, versioning makes this even more complex which, in my mind, reinforces the need for a consistent "root" name. If we wanted to experiment more with the nickname approach, perhaps something like this would work:

Specification: LAPIS, v1.0
XML encoding: (LAPIS-X1, XACML4)
JSON encoding: (LAPIS-J1)
YAML encoding: (LAPIS-Y1)

...where the nicknames refer to an explicit major version: LAPIS-Y1 = LAPIS, v1.0 expressed in YAML. How are minor versions handled? I have some ideas, but was hoping we could get some of traction on the base level naming before heading down that rabbit hole. 

However this plays out I think we need to agree upon the *before* making large scale announcements so that we not only have consistent messaging but also, so that we can begin the process of associating NextGen work with something permanent.

**Comment by steven-legg** [comment]
> I spent some time exploring names with the LLMs to see if I could come up with a name for the core specification and this one started to grow on me:
> 
> **LAPIS: Language-Agnostic Policy Interchange Specification**

I can live with that.

> This brings me back to the conundrum of naming the various output formats. I agree that we need to find a way to retain “XACML”, but at the same time I don’t think we want to create a bag of names due to the risk of confusion. Building upon the output above I was thinking that maybe a way to do this would be to create a base set of names and then provide a “nickname” for the XML version (something that is common in things like Linux versioning). In that situation the spec naming could look something like this:
> 
> Specification: LAPIS, v1.0
> XML encoding: LAPIS-X v1.0 (XACML4)
> JSON encoding: LAPIS-J, v1.0
> YAML encoding: LAPIS-Y, v1.0
> … 
> ALFA encoding: LAPIS-A, v1.0 (ALFA2) ...?

ALFA 2.0 is likely to be a subset of LAPIS functionality so it wouldn't rightly have a LAPIS-* name.

> As discussed earlier, versioning makes this even more complex which, in my mind, reinforces the need for a consistent "root" name. If we wanted to experiment more with the nickname approach, perhaps something like this would work:
> 
> Specification: LAPIS, v1.0
> XML encoding: (LAPIS-X1, XACML4)
> JSON encoding: (LAPIS-J1)
> YAML encoding: (LAPIS-Y1)

I was fine with LAPIS-J, v1.0, etc. The exact version is usually established in the introduction or normative references. The name without a version number is commonly used otherwise.

> ...where the nicknames refer to an explicit major version: LAPIS-Y1 = LAPIS, v1.0 expressed in YAML. How are minor versions handled? I have some ideas, but was hoping we could get some of traction on the base level naming before heading down that rabbit hole.

A change to the core specification is likely to have knock-on changes to the representations. So align the major and minor version numbers across the specifications. Where a representation, or the core, gets an update without affecting the semantics then increment the sub-minor number. So LAPIS 1.0.1, LAPIS-X 1.0, LAPIS-J 1.0.3 belong together because they all start with 1.0.

Or reserve the major version number changes for semantic changes that affect the whole suite, even if fairly minor, and use minor number changes for changes that affect only that specification.

**Comment by humantypo** [comment]
@cdanger , @Hal-Lockhart any thoughts on this? I was hoping that we might be able to introduce some element of the naming in the the blog post I am working with Oasis to "announce" the v4 work.

**Comment by steven-legg** [comment]
Acronym-Resistant Policy Language (ARPL) 😄
